src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerActor.scala

```diff
@@ -266,13 +268,30 @@ private[impl] class GroupManagerActor(
 
     val dynamicApps: Set[AppDefinition] =
       to.transitiveApps.map {
-        // assign values for service ports that the user has left "blank" (set to zero)
-        case app: AppDefinition if app.hasDynamicServicePorts => assignPorts(app)
-        case app: AppDefinition =>
-          // Always set the ports to service ports, even if we do not have dynamic ports in our port mappings
+        case app: AppDefinition if app.servicePorts.nonEmpty =>
+          synchronized {
+            app.servicePorts.withFilter(_ != AppDefinition.RandomPortValue).foreach { port => taken += port }
+          }
+          assignPorts(app)
+        case app: AppDefinition if app.portDefinitions.exists(_.port == AppDefinition.RandomPortValue) =>
```

Comment:
As stated, I'm reverting the change as is and reverting service port global
uniqueness.

There is absolutely no reason to have globally unique ports: even poor mans
load balancing has to consider which hosts have a given app, so if another
app is using the same port, they simply can't be collocated. I have no
intention of breaking compatibility, but this "global service port" change
is absolutely a regression and compatibility breaking. It's also why the
bug you pointed out with unrelated deployments being blocked on each other
occurs.
On Tue, Aug 9, 2016 at 9:45 AM Gastón Kleiman <notifications@github.com>
wrote:

> In
> src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerActor.scala
> <https://github.com/mesosphere/marathon/pull/4195#discussion_r74096669>:
>
> > @@ -266,13 +268,30 @@ private[impl] class GroupManagerActor(
> >
> >      val dynamicApps: Set[AppDefinition] =
> >        to.transitiveApps.map {
> > -        // assign values for service ports that the user has left "blank" (set to zero)
> > -        case app: AppDefinition if app.hasDynamicServicePorts => assignPorts(app)
> > -        case app: AppDefinition =>
> > -          // Always set the ports to service ports, even if we do not have dynamic ports in our port mappings
> > +        case app: AppDefinition if app.servicePorts.nonEmpty =>
> > +          synchronized {
> > +            app.servicePorts.withFilter(_ != AppDefinition.RandomPortValue).foreach { port => taken += port }
> > +          }
> > +          assignPorts(app)
> > +        case app: AppDefinition if app.portDefinitions.exists(_.port == AppDefinition.RandomPortValue) =>
>
> portDefinitions were supposed to be a superset of ports. I think that we
> shouldn't make breaking changes of the interface, without having clearly
> communicated a deprecation plan.
>
> Until recently, our docs used to say that the user was responsibly for
> ensuring the uniqueness of the service ports, and there was no validation.
>
> Then the docs were updated to say that Marathon enforces the uniqueness,
> and a broken validation was added.
>
> I recently fixed the validation. But I'd be ok with going back to not
> enforcing the uniqueness, as long as that is properly documented and maybe
> even checked with the community.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mesosphere/marathon/pull/4195/files/a4f3676c25313e8b30b754ba3f0fc7c9356b234f#r74096669>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAweZqUjMj5SUroJgDleiXGWwbnzYZ8Lks5qeK6ngaJpZM4JfdWT>
> .
>